### PR TITLE
chore: update to `react-router-dom` v6

### PR DESCRIPTION
**Pull request by keita-determined**
## Description
[WEB-172]
[WEB-518]

Previous: #5166
Followed [the instruction](https://github.com/remix-run/react-router/discussions/8753)

What I have done in this PR
- Update `react-router-dom`, `@types/react-router-dom`, and `history` npm libraries
- Replace `react-router-dom-v5-compat` with `react-router-dom`
- Replace `Prompt` cuz its not available in `react-router-dom v6`
  - Fixed bugs [WEB-518]

## Plan

```mermaid
gantt
    title Update React-Router-Dom Plan
    dateFormat  YYYY-MM-DD
    section React-Router v6
    Replace hooks                           :a1, 2022-09-28, 3d
    Replace Root                             :a2, after a1 , 5d
    Wrap up React Router v6         :crit, a3, after a2 , 5d
    section React v18
    Update to React v18      :b1,  2022-10-12, 5d
```

## Test Plan
- Check if all routes works as expected
- Check if prompt shows up if user is still editing `Note` 

## Commentary (optional)
- Lots of files are changed, but mostly renaming `react-router-dom-v5-compat` to `react-router-dom`

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

[WEB-172]: https://determinedai.atlassian.net/browse/WEB-172?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[WEB-518]: https://determinedai.atlassian.net/browse/WEB-518?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/ee/SamlAuth.test.ts [hunk 1]
```diff
@@ -51,11 +49,5 @@ describe('SamlAuth', () => {
     it('should decode and flatten relayState query param', () => {
       expect(utils.handleRelayState(QUERIES_INPUT)).toStrictEqual(QUERIES_OUTPUT);
     });
-
-    it('should add to history', () => {
-      const historyLength = history.length;
```
Comment (keita-determined) [inline webui/react/src/ee/SamlAuth.test.ts:56]: `history v5` doesn't have length, so this test case isn't valid

File: webui/react/src/components/NotesCard.tsx [hunk 2]
```diff
@@ -52,6 +52,31 @@ const NotesCard: React.FC<Props> = ({
     setEditedNotes(existingNotes.current);
   }, [noteChangeSignal]);
 
+  useEffect(() => {
```
Comment (keita-determined) [inline webui/react/src/components/NotesCard.tsx:55]: alternative of Prompt cuz new `react-router-dom` doesnt have `prompt` yet

File: webui/react/src/index.tsx [hunk 3]
```diff
@@ -1,7 +1,11 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
-import { Router } from 'react-router-dom';
-import { CompatRouter } from 'react-router-dom-v5-compat';
+/**
+ * It's considered unstable until `react-router-dom` can detect
+ * history version mismatches when supplying your own history.
+ * https://reactrouter.com/docs/en/v6/routers/history-router
+ */
+import { unstable_HistoryRouter as HistoryRouter } from 'react-router-dom';
```
Comment (keita-determined) [inline webui/react/src/index.tsx:8]: This should be safe cuz the version of `history` in `react-router-dom` is the same as `history` we use

File: webui/react/src/index.tsx [hunk 4]
```diff
@@ -1,7 +1,11 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
-import { Router } from 'react-router-dom';
-import { CompatRouter } from 'react-router-dom-v5-compat';
+/**
+ * It's considered unstable until `react-router-dom` can detect
+ * history version mismatches when supplying your own history.
+ * https://reactrouter.com/en/v6.3.0/api#unstable_historyrouter
+ */
+import { unstable_HistoryRouter as HistoryRouter } from 'react-router-dom';
```
Comment (keita-determined) [inline webui/react/src/index.tsx:8]: This should be safe cuz the version of `history` in `react-router-dom` is the same as `history` we use

File: webui/react/src/shared/routes/history.ts [hunk 5]
```diff
@@ -1,2 +1,2 @@
 import { createBrowserHistory } from 'history';
-export default createBrowserHistory({ basename: process.env.PUBLIC_URL });
+export default createBrowserHistory();
```
Comment (keita-determined) [inline webui/react/src/shared/routes/history.ts:2]: new `history` doesnt take arguments anymore. `basename` is moved to `index.tsx`

File: webui/react/src/components/ThemeToggle.test.tsx [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import { render, screen } from '@testing-library/react';
 import userEvent from '@testing-library/user-event';
 import React from 'react';
-import { Router } from 'react-router-dom';
+import { unstable_HistoryRouter as HistoryRouter } from 'react-router-dom';
```
Comment (mapmeld) [inline webui/react/src/components/ThemeToggle.test.tsx:4]: is there a documentation URL about this? would be good to make a comment or know if this is being deprecated

Comment (keita-determined) [reply webui/react/src/components/ThemeToggle.test.tsx:4]: Yes, https://reactrouter.com/en/v6.3.0/api#unstable_historyrouter
We need to use this to use `history` lib

File: webui/react/src/ee/SamlAuth.test.ts [hunk 7]
```diff
@@ -51,11 +49,5 @@ describe('SamlAuth', () => {
...
-      utils.handleRelayState(QUERIES_INPUT);
-      expect(history).toHaveLength(historyLength + 1);
```
[the remark this answers is not in the document]
Comment (keita-determined) [reply webui/react/src/ee/SamlAuth.test.ts:58]: We can do it like https://github.com/remix-run/history/blob/dev/docs/api-reference.md#navigation

File: webui/react/src/components/NotesCard.tsx (hunk 2 above)
Comment (mapmeld) [reply webui/react/src/components/NotesCard.tsx:55]: ✅  tested editing note and getting ok/cancel out

**Comment by mapmeld** [comment]
LGTM ✅

**Review by mapmeld — APPROVED** [review]

**Pull request merged**